sdks/go/pkg/beam/io/datastoreio/datastore.go

```diff
@@ -32,8 +32,20 @@ import (
 	"github.com/apache/beam/sdks/v2/go/pkg/beam/internal/errors"
 	"github.com/apache/beam/sdks/v2/go/pkg/beam/log"
 	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
 )
 
+type clientType interface {
+	Run(context.Context, *datastore.Query) *datastore.Iterator
+	Close() error
+}
+
+// call to newClient returns an instance of datastore.Client
+// redirect this call for unit test usage
+var newClient = func(ctx context.Context, projectID string, opts ...option.ClientOption) (clientType, error) {
```

Comment:
The problem with this approach is that it affects a global/package level value, preventing parallel test execution if that's something we'd like to do in the future. 

I would recommend the following approach instead:
* modifying the unexported `query` function to take in a function parameter of the signature of this newClient.
* In the production calls (`Read`) pass a nil in that position.
* In the DoFns that require the newClient call add an unexported field of the newClient function type
* Populate that field with the parameter.
* To the affected DoFns, add a Setup() function, that looks at that newClient field, and if it's nil, set it to the datastore calling version of the function (preferably defined somewhere as a named function, not as a variable). Otherwise, don't touch it.
* in the ProcessElement methods, call the function field.

This will let us override the newClient function in tests, at least when using the direct runner. It will also avoid the package level override, and won't interrupt parallel execution of the tests at a later time. No package level override also avoids needing to set and reset the field.

It won't fix things for portablee runners, as that would require a bit more effort. These tests are currently only run on the direct runner, so that's fine. This is for unit testing.

Reply:
Thanks so much for the suggestions! If I understood correctly, does the unit test then have to call the unexported `query` function directly, with a newClient function type?